### ISSUE TITLE
PARTITION BY doesn't work for dot-notated columns (i.e. columns of aliased objects)

### ISSUE THREAD

**Issue by rmoff** [confluentinc/ksql#1475]
Good

```
ksql> CREATE STREAM TEST01 AS SELECT R.MESSAGE, C.LAST_NAME FROM RATINGS R LEFT JOIN CUSTOMERS C \
> ON R.USER_ID = C.ID PARTITION BY LAST_NAME;

 Message
----------------------------
 Stream created and running
----------------------------
```

Bad

```
ksql> CREATE STREAM TEST01 AS SELECT R.MESSAGE, C.LAST_NAME FROM RATINGS R LEFT JOIN CUSTOMERS C \
> ON R.USER_ID = C.ID PARTITION BY C.LAST_NAME;
line 1:126: mismatched input '.' expecting ';'
Caused by: org.antlr.v4.runtime.InputMismatchException
```

This is a problem if there's an ambiguous reference (e.g. both objects in the JOIN had a column called `LAST_NAME`. In that instance you'd have to alias one of them, and partition on that: 

```
ksql> CREATE STREAM TEST02 AS SELECT R.MESSAGE, C.LAST_NAME AS "FOO" FROM RATINGS R LEFT JOIN CUSTOMERS C \
> ON R.USER_ID = C.ID PARTITION BY FOO;

 Message
----------------------------
 Stream created and running
----------------------------
```

**Comment by jzaralim** [comment]
Closing because this is expected behavior.

The `PARTITION BY` clause should only accept column names in the query result and not from the sources. For example,

`CREATE STREAM TEST AS SELECT RATINGS.MESSAGE, RATINGS.USER_ID FROM RATINGS PARTITION BY RATINGS.USER_ID;`

Fails because `RATINGS.USER_ID` refers to a column in the source rather than the resulting stream. The name of the column in the result is actually `USER_ID`.

In the case of ambiguous references, if the user doesn’t provide an alias, KSQL will automatically generate one. For example. the columns in the result of the following query are `MESSAGE` and `C_LAST_NAME`.

`CREATE STREAM TEST01 AS SELECT R.MESSAGE, C.LAST_NAME FROM RATINGS R LEFT JOIN CUSTOMERS C  ON R.USER_ID = C.ID PARTITION BY C.LAST_NAME;`

In this case, the `PARTITION BY` clause should use `C_LAST_NAME` instead.

**Comment by rmoff** [comment]
> Closing because this is expected behavior.

Expected by whom? :-) 

I'd argue that we oughtn't "doctape" an issue, but look at it from the point of view of the end user. It's not intuitive to look at it this way. Can we consider re-opening this and having KSQL actually handle it in a more "SQL"y way, i.e. consistent use of column naming?

**Comment by MichaelDrogalis** [comment]
I agree with @rmoff. This restriction feels like an artifact of the implementation. It's not something we should consciously carry through to the end user.

**Comment by TroyCoombs** [comment]
I assume this issue is still legit, as I'm encountering it with the following CREATE STREAM,
```
CREATE STREAM EVENTS_API_SEARCH
      WITH (KAFKA_TOPIC='events-for-search-api') AS
SELECT E.AFTER->ID as id,
      E.AFTER->TEAM_ID as team_id,
      E.AFTER->CREATED_AT AS created_at,
      E.AFTER->CLI_ID AS cli_id,
      C.USER_ID AS cli_user_id
FROM    EVENTS E
      LEFT JOIN  CLI_CLIENTS_MODEL_BY_ID C
      ON COALESCE(E.AFTER->CLI_ID, '') = C.ID
PARTITION BY E.AFTER->ID;
```

I can run this via ksql cli just fine, but I get the error as mentioned by @rmoff  when it runs as part of a script during a ksqldb-cli docker container command.  Which I don't understand???

I did try running it manually by changing the `PARTITION BY` to be 'id' (E.AFTER->ID alias) but I had to include C.ID in the projection, and then I get weird results where the values for "E.AFTER->ID" actually is projected in the 'cli_id2' field.  Which also confuses me.
```
CREATE STREAM HOOKDECK_CORE_EVENTS_API_SEARCH
      WITH (KAFKA_TOPIC='events-for-search-api-search') AS
SELECT E.AFTER->ID as id,
      E.AFTER->TEAM_ID as team_id,
      E.AFTER->CLI_ID AS cli_id,
      E.AFTER->_STATUS AS _status,
      E.AFTER->STATE AS state,
      C.USER_ID AS cli_user_id,
      C.ID as cli_id2
FROM    HOOKDECK_CORE_EVENTS E
      LEFT JOIN  HOOKDECK_CORE_CLI_CLIENTS_MODEL_BY_ID C
      ON COALESCE(E.AFTER->CLI_ID, '') = C.ID
PARTITION BY id;
```

Note:  I'm fairly new to the kafka world so some of this could be a beginner that just needs a little guidance.
